### DISCUSSION TITLE
Authentication / Authorization on custom endpoints

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#8336] [Archive]
Hi all,

Just wondering, what are the best practices for authentication and authorization on custom endpoints in Directus? And are there any examples? I'm looking for a way to keep user management and roles in the app, but also applying that logic to custom endpoints.

Cheers!

Kees.

**Comment by joselcvarela** [comment]
Hey @keesvanbemmel,
Here you have https://docs.directus.io/guides/api-endpoints/#custom-api-endpoints

So when you use a service, you need to pass `req.accountability`. This object keeps the data about the current logged user.
So if current user is not authorized to make some operation, a 401 response will be received.

**Reply by keesvanbemmel** [reply]
Thanks!

But I guess my question is more: where should I put the logic that the user (from req.accountability) can indeed access that specific custom endpoint?

**Reply by rijkvanzanten** [reply]
That'd be in the custom endpoint itself, there's not "permissions configuration" for custom endpoints from within the app 👍🏻

**Reply by keesvanbemmel** [reply]
Ok, perhaps we'll create an endpoint collection then and add roles/users to it! thanks!

**Reply by rijkvanzanten** [reply]
Yeah! That'd be a great solution 🙂

**Reply by keesvanbemmel** [reply]
@rijkvanzanten any experience with setting custom access rules based on current user role?

I'm trying this:
`{
    "roles": {
        "name": {
            "_contains": "$CURRENT_ROLE"
        }
    }
}`

Created a collections called "endpoints" with a m2m field called "roles" linking to directus system roles.

**Reply by rijkvanzanten** [reply]
@keesvanbemmel The "$CURRENT_ROLE" flag is the UUID of the role, not the name. 

Seeing this is a m2m, I'm assuming you have a junction table in between as well, so it's more like 

```
current                    junction               directus_roles
---------                  ---------              ---------------
                           id
id            <------      current_id
                           role_id     ----->     id
roles (o2m to junction)
```

So in your rule, try using

```json
{
  "roles": {
    "role_id": {
      "_eq": "$CURRENT_ROLE"
    }
  }
}
```

**Reply by keesvanbemmel** [reply]
Hero!

One quick follow up: will it iterate over all linked roles and check if at least one is _eq?

**Reply by rijkvanzanten** [reply]
> will it iterate over all linked roles and check if at least one is _eq?

Yes. It's a "one of" sorta deal, not a "all of them" 👍🏻

**Reply by keesvanbemmel** [reply]
I just added (linked) two roles. One of which is the current role of the user, but the _eq filter doesn't work there.

**Reply by keesvanbemmel** [reply]
So I thought this would've worked:

`{
    "roles": {
        "directus_roles_id": {
            "_contains": "$CURRENT_ROLE"
        }
    }
}`

Seems like the _contains operator is just for substrings, not array items. Any thoughts?

EDIT:
Interestingly, this seems to work:
`{
    "roles": {
        "directus_roles_id": {
            "_in": "$CURRENT_ROLE"
        }
    }
}`

Even though the _in operator seems "backwards", this renders true if the CURRENT_ROLE is in the list of directus_roles_id's

**Reply by rijkvanzanten** [reply]
> Seems like the _contains operator is just for substrings, not array items. Any thoughts?

This is correct. _contains uses SQL `LIKE %{substring}%`, so that doesn't work on arrays. 

The `_in` uses SQL's `IN (a, b, c)`, which is effectively "Equals A, or equals B, or equals C", and that's done for every item in the array 🤔

**Reply by keesvanbemmel** [reply]
So you're saying _eq should work in this case even if two or more roles are linked?

**Reply by rijkvanzanten** [reply]
Yes. Filtering the top level items based on a filter that looks at a nested one to many set works on a "has one of the following" basis
